File: java/xss/TemplateInjection.java
import freemarker.template.*;
import jakarta.servlet.http.*;
import org.apache.velocity.VelocityContext;
import org.apache.velocity.app.VelocityEngine;
import org.apache.velocity.runtime.RuntimeConstants;
import org.stringtemplate.v4.*;
import java.io.*;
import java.util.Map;

public class TemplateInjection extends HttpServlet {

    // VULN 1: Freemarker template injection - user controls the template string directly
    public String renderEmailTemplate(HttpServletRequest req, Map<String, Object> model) throws Exception {
        String userTemplate = req.getParameter("email_template");
        Configuration cfg = new Configuration(Configuration.VERSION_2_3_32);
        cfg.setClassForTemplateLoading(TemplateInjection.class, "/templates");
        Template template = new Template("userTemplate", new StringReader(userTemplate), cfg);
        StringWriter out = new StringWriter();
        template.process(model, out);
        return out.toString();
    }

    // VULN 2: Velocity template injection - user controls the template string for order receipts
    public String renderOrderReceipt(HttpServletRequest req, Map<String, Object> orderData) {
        String templateStr = req.getParameter("receipt_template");
        VelocityEngine ve = new VelocityEngine();
        ve.setProperty(RuntimeConstants.RESOURCE_LOADER, "string");
        ve.addProperty("string.resource.loader.class",
            "org.apache.velocity.runtime.resource.loader.StringResourceLoader");
        ve.init();
        VelocityContext ctx = new VelocityContext(orderData);
        StringWriter writer = new StringWriter();
        ve.evaluate(ctx, writer, "receipt", templateStr);
        return writer.toString();
    }

    // VULN 3: StringTemplate (SSTI) with user-controlled template content for product descriptions
    public String renderProductDescription(HttpServletRequest req) {
        String userContent = req.getParameter("description_template");
        STGroupString group = new STGroupString("shopist", userContent, '$', '$');
        ST st = group.getInstanceOf("product_desc");
        if (st == null) {
            st = new ST(userContent, '$', '$');
        }
        st.add("siteName", "Shopist");
        return st.render();
    }
}

File: kotlin/xss/TemplateInjection.kt
import freemarker.template.Configuration
import freemarker.template.Template
import jakarta.servlet.http.HttpServletResponse
import org.apache.velocity.VelocityContext
import org.apache.velocity.app.VelocityEngine
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.io.StringReader
import java.io.StringWriter

data class TemplateRequest(val templateSource: String, val productName: String, val price: Double)

@RestController
class TemplateInjection {

    // VULN 1: Freemarker template loaded directly from user-supplied string - custom email templates
    @PostMapping("/api/marketing/preview-email")
    fun previewEmailTemplate(@RequestBody request: TemplateRequest, response: HttpServletResponse): String {
        val cfg = Configuration(Configuration.VERSION_2_3_32)
        val template = Template("user-template", StringReader(request.templateSource), cfg)
        val model = mapOf(
            "productName" to request.productName,
            "price" to request.price,
            "shopName" to "Shopist"
        )
        val writer = StringWriter()
        template.process(model, writer)
        return writer.toString()
    }

    // VULN 2: Velocity engine rendering user-controlled template - product description renderer
    @PostMapping("/api/products/render-description")
    fun renderProductDescription(@RequestBody request: TemplateRequest): String {
        val engine = VelocityEngine()
        engine.init()
        val context = VelocityContext()
        context.put("productName", request.productName)
        context.put("price", request.price)
        val writer = StringWriter()
        engine.evaluate(context, writer, "product-desc", request.templateSource)
        return writer.toString()
    }

    // VULN 3: Thymeleaf processing user-controlled template expression - storefront customization
    @PostMapping("/api/storefront/custom-banner")
    fun renderCustomBanner(
        @RequestParam("templateExpr") templateExpr: String,
        @RequestParam("shopName") shopName: String,
        response: HttpServletResponse
    ): String {
        val org = org.thymeleaf.TemplateEngine()
        val resolver = org.thymeleaf.templateresolver.StringTemplateResolver()
        org.setTemplateResolver(resolver)
        val ctx = org.thymeleaf.context.Context()
        ctx.setVariable("shopName", shopName)
        ctx.setVariable("expr", templateExpr)
        val fullTemplate = "<div th:utext=\"${templateExpr}\">Banner</div>"
        return org.process(fullTemplate, ctx)
    }
}
